### PR TITLE
Adding plots for the hltbtag HEP17 validation

### PR DESCRIPTION
**Pull request by natalia-korneeva**
Adding the usual validation plots for specific modules (HEP17, HEM17, HEP18).
Adding the HEP17/HEM17 relation plots.
An example of relation plots are here:
https://twiki.cern.ch/twiki/bin/view/Sandbox/NataliaTsirovaSandbox#Relations_HEP17_HEM17

### CODE REVIEW

**Comment by heppye** [comment]
Greenlight from BTV@HLT. Please start the test. Thanks, Ye

File: HLTriggerOffline/Btag/src/HLTBTagPerformanceAnalyzer.cc [hunk 1]
```diff
@@ -139,8 +141,17 @@ void HLTBTagPerformanceAnalyzer::analyze(const edm::Event& iEvent, const edm::Ev
 		}
 
 		for (auto & BtagJT: JetTag) {
+			std::map<std::string, bool> inmodule;
+			inmodule["HEP17"]=(BtagJT.first->phi() >= -0.87) && (BtagJT.first->phi() < -0.52) && (BtagJT.first->eta() > 1.3);
```
Comment (mtosi) [inline HLTriggerOffline/Btag/src/HLTBTagPerformanceAnalyzer.cc]: wouldn't it be better to have a function for handling the HE slices ?
(instead of having these number hard coded)

Comment (natalia-korneeva) [inline HLTriggerOffline/Btag/src/HLTBTagPerformanceAnalyzer.cc]: @mtosi we need to monitor only 3 slices (HEP17, HEM17, HEP18), so any additional function will increase the length of the code. I'm not sure it is necessary in this moment. Or do you mean something else?

**Comment by mtosi** [comment]
I think it would make less error prone
  mia

**Comment by natalia-korneeva** [comment]
@mtosi I got your point and I agree. But since now there are only 3 slices I would leave it as is. If it would be necessary to monitor more modules then of course such changes would be introduced.

File: HLTriggerOffline/Btag/src/HLTBTagHarvestingAnalyzer.cc [hunk 2]
```diff
@@ -163,6 +206,14 @@ bool HLTBTagHarvestingAnalyzer::GetNumDenumerators(DQMStore::IBooker& ibooker, D
 		delete cutg_num;
 		delete cutg_den;
 	}
+
+	if (type==3) //mistagrate HEP17 / HEM17 relation: fill "ptrnum" with HEP17 mistagrate and "ptrden" with HEM17 mistagrate
+	{
+		TH1* numH1 = numME->getTH1();
+		TH1* denH1 = denME->getTH1();
+		ptrden=(TH1*)denH1->Clone("denominator");
```
Comment (dmitrijus) [inline HLTriggerOffline/Btag/src/HLTBTagHarvestingAnalyzer.cc]: These pointers are created, but as far as I see, not deleted.
Please make sure you delete them!

And please avoid "return by modifying a reference".
`TH1 * & ptrnum` is just horrible, no matter how you look at it :(

Use std::tuple to return multiple values instead.
`std::tuple<bool, TH1*, TH1*> HLTBTagHarvestingAnalyzer::GetNumDenumerators`

Comment (dmitrijus) [inline HLTriggerOffline/Btag/src/HLTBTagHarvestingAnalyzer.cc]: These pointers are created, but as far as I see, not deleted.
Please make sure you delete them!

And please avoid "return by modifying a reference".
`TH1 * & ptrnum` is just horrible, no matter how you look at it :(

Use std::tuple to return multiple values instead.
`std::tuple<bool, TH1*, TH1*> HLTBTagHarvestingAnalyzer::GetNumDenumerators`

**Review by dmitrijus — CHANGES_REQUESTED** [review]

File: HLTriggerOffline/Btag/src/HLTBTagHarvestingAnalyzer.cc (hunk 2 above)
Comment (fwyzard) [inline HLTriggerOffline/Btag/src/HLTBTagHarvestingAnalyzer.cc]: One should *never* delete the pointers returned by `getTH1()` as they point to the objects owned by the DQM service.

Comment (fwyzard) [inline HLTriggerOffline/Btag/src/HLTBTagHarvestingAnalyzer.cc]: {quote}
And please avoid "return by modifying a reference".
{quote}
I'm reasonably sure you can find arguments going both ways here - it mostly depends on one's background (e.g. C vs python).

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
what's holding this up ?

File: HLTriggerOffline/Btag/src/HLTBTagHarvestingAnalyzer.cc (hunk 2 above)
Comment (natalia-korneeva) [inline HLTriggerOffline/Btag/src/HLTBTagHarvestingAnalyzer.cc]: @dmitrijus I removed the unnecessary copy.
Concerning "return by modifying a reference": I may agree with you, but this is out of scope of the current PR. So to avoid any confusion I would leave this improvement to a separate PR.

**Comment by dmitrijus** [comment]
> One should never delete the pointers returned by getTH1() as they point to the objects owned by the DQM service.

@fwyzard Correct. But you must delete your own pointers, created via new, malloc, or in this case, Clone(). This PR (still) creates a few such pointers.

And this was a reason why I wanted a more clear function interface. In current one, it is not clear if the new pointers are created or not.

@natalia-korneeva Please do ctrl+f on Clone and make sure each Clone is properly erased.

**Comment by dmitrijus** [comment]
+1

regarding my last comment, the PR adds no additional leaks.
Clone() is still misused in the file this PR touches and should be eventually fixed, via separate PR, eg. [1](https://github.com/natalia-korneeva/cmssw/blob/315b77d12e6136cff9799f36e5cc3beada734806/HLTriggerOffline/Btag/src/HLTBTagHarvestingAnalyzer.cc#L151)

File: HLTriggerOffline/Btag/interface/HLTBTagHarvestingAnalyzer.h [hunk 3]
```diff
@@ -48,6 +49,8 @@ class HLTBTagHarvestingAnalyzer : public DQMEDHarvester {
 			// Histogram handler
 			std::map<std::string, MonitorElement *> H1_;
 
+			std::vector<std::string> modules_;
```
Comment (davidlange6) [inline HLTriggerOffline/Btag/interface/HLTBTagHarvestingAnalyzer.h]: please make this a vector of enums. That will save a lot of string comparisons.

File: HLTriggerOffline/Btag/src/HLTBTagHarvestingAnalyzer.cc [hunk 4]
```diff
@@ -27,11 +29,14 @@ HLTBTagHarvestingAnalyzer::dqmEndJob(DQMStore::IBooker & ibooker, DQMStore::IGet
 	{
 		dqmFolder_hist = Form("HLT/BTag/Discriminator/%s",hltPathNames_[ind].c_str());
 		std::string effDir = Form("HLT/BTag/Discriminator/%s/efficiency",hltPathNames_[ind].c_str());
+		std::string relationsDir = Form("HLT/BTag/Discriminator/%s/HEP17_HEM17",hltPathNames_[ind].c_str());
 		ibooker.setCurrentFolder(effDir);
 		TH1 *den =NULL;
 		TH1 *num =NULL; 
 		std::map<TString,TH1F> effics;
 		std::map<TString,bool> efficsOK;
+		std::map<TString,std::map<std::string,TH1F> > efficsmod;
```
Comment (davidlange6) [inline HLTriggerOffline/Btag/src/HLTBTagHarvestingAnalyzer.cc]: if a string is really needed here, it should be a std::string not a TString

File: HLTriggerOffline/Btag/interface/HLTBTagHarvestingAnalyzer.h (hunk 3 above)
Comment (natalia-korneeva) [inline HLTriggerOffline/Btag/interface/HLTBTagHarvestingAnalyzer.h]: I need a string to construct the path, for example:
https://github.com/natalia-korneeva/cmssw/blob/HEP17-btaghlt-validation/HLTriggerOffline/Btag/src/HLTBTagHarvestingAnalyzer.cc#L55

File: HLTriggerOffline/Btag/src/HLTBTagHarvestingAnalyzer.cc (hunk 4 above)
Comment (natalia-korneeva) [inline HLTriggerOffline/Btag/src/HLTBTagHarvestingAnalyzer.cc]: Could you please comment more, why it should be std::string?
The first key is of a TString type:
https://github.com/natalia-korneeva/cmssw/blob/HEP17-btaghlt-validation/HLTriggerOffline/Btag/src/HLTBTagHarvestingAnalyzer.cc#L60
https://github.com/natalia-korneeva/cmssw/blob/HEP17-btaghlt-validation/HLTriggerOffline/Btag/src/HLTBTagHarvestingAnalyzer.cc#L44

Comment (fwyzard) [inline HLTriggerOffline/Btag/src/HLTBTagHarvestingAnalyzer.cc]: I think the request is to use `std::string` instead of `TString`` everywhere.
For what is worth, I agree.

Comment (natalia-korneeva) [inline HLTriggerOffline/Btag/src/HLTBTagHarvestingAnalyzer.cc]: @fwyzard well yes, it is better to use std::string instead of TString, but it is out of scope of the current PR. The purpose of this one is not to improve the current code, but just to add the necessary HEP17 plots.
To my mind it's better to separate these two tasks.

Comment (fwyzard) [inline HLTriggerOffline/Btag/src/HLTBTagHarvestingAnalyzer.cc]: fine by me, but please do prepare a follow up PR for that.

**Comment by fwyzard** [comment]
To be clear

> fine by me, but please do prepare a follow up PR for that.

means to do that by the June 20th deadline, in order to have the monitored triggers deployed online.

**Comment by fwyzard** [comment]
Sorry, I though this was for new paths, instead I see it is for monitoring existing objects.

Please make a follow up PR fixing
  - proper handling of Cloned histograms
  - usege of TString

File: HLTriggerOffline/Btag/src/HLTBTagHarvestingAnalyzer.cc (hunk 4 above)
Comment (natalia-korneeva) [inline HLTriggerOffline/Btag/src/HLTBTagHarvestingAnalyzer.cc]: @fwyzard OK, I will. Then I would leave the current one as is and introduce the mentioned improvements (cloned histograms and TString) in a separate PR.

File: HLTriggerOffline/Btag/interface/HLTBTagHarvestingAnalyzer.h (hunk 3 above)
Comment (davidlange6) [inline HLTriggerOffline/Btag/interface/HLTBTagHarvestingAnalyzer.h]: is modules not just

modules_ = {"HEP17", "HEP18", "HEM17"};

File: HLTriggerOffline/Btag/src/HLTBTagHarvestingAnalyzer.cc (hunk 4 above)
Comment (davidlange6) [inline HLTriggerOffline/Btag/src/HLTBTagHarvestingAnalyzer.cc]: indeed - it at least looks trivial to change the key for this new string, as the TString key for the existing maps is just a TString converted from a std::string.

File: HLTriggerOffline/Btag/interface/HLTBTagHarvestingAnalyzer.h (hunk 3 above)
Comment (natalia-korneeva) [inline HLTriggerOffline/Btag/interface/HLTBTagHarvestingAnalyzer.h]: Right, we agree to monitor the plots in these three slices. The histograms are filled and put to the corresponding folder named "HEP17" etc.

File: HLTriggerOffline/Btag/src/HLTBTagHarvestingAnalyzer.cc (hunk 4 above)
Comment (natalia-korneeva) [inline HLTriggerOffline/Btag/src/HLTBTagHarvestingAnalyzer.cc]: @davidlange6 Done.

**Comment by dmitrijus** [comment]
+1

File: HLTriggerOffline/Btag/interface/HLTBTagHarvestingAnalyzer.h (hunk 3 above)
Comment (davidlange6) [inline HLTriggerOffline/Btag/interface/HLTBTagHarvestingAnalyzer.h]: hi @natalia-korneeva - sorry I missed your answer - I see your point now - so the needed change to make everything work is to make a small map:
enum HCALSpecials = { HEP17, HEP18, HEM17};
std::map<HCALSpecials,std::string> hcalSpecialNames = { HEP17: "HEP17", HEP18: "HEP18", HEM17: "HEM17" };

and then use the enums to index and look up the histograms, and the map to get from enum to string that you need for constructing the histograms.

Does that make sense?

Comment (natalia-korneeva) [inline HLTriggerOffline/Btag/interface/HLTBTagHarvestingAnalyzer.h]: Yes, but in this file the variable modules_ is used in the dqmEndJob function only, so it is called only once per job and I guess this change won't improve performance

Comment (davidlange6) [inline HLTriggerOffline/Btag/interface/HLTBTagHarvestingAnalyzer.h]: hi @natalia-korneeva - sorry i missed your answer - i was worried instead about modules_ in HLTBTagPerformanceAnalyzer::analyze

Comment (natalia-korneeva) [inline HLTriggerOffline/Btag/interface/HLTBTagHarvestingAnalyzer.h]: Thanks for the clarification, I understand your worries. I will work on this.

Comment (natalia-korneeva) [inline HLTriggerOffline/Btag/interface/HLTBTagHarvestingAnalyzer.h]: @davidlange6 done.

**Comment by fwyzard** [comment]
@cmsbuild please test

**Comment by davidlange6** [comment]
merge

**Comment by natalia-korneeva** [comment]
@fwyzard I've created a follow up PR fixing a TString usage
https://github.com/cms-sw/cmssw/pull/20549

Concerning the handling of Cloned histograms could you please comment more on improvements: looking at the code I don't see any problems.

**Comment by fwyzard** [comment]
@natalia-korneeva here are a couple of general comments about the usage of `Clone()`, and in general the usage of `new` and `delete`:

  - when you create a new object (e.g. TH1F) in a method and delete it at the end of the same method, simply create it as a local variable: e.g. `TH1F eff(*num);` rather than `TH1F* eff=new TH1F(*num);`; this way you don't need to `delete` it at the end of the method or of the block.

  - when you create a new object with `Clone(...)` or `ProjectionY(...)`, it should be deleted at some point.

About the specific case of `GetNumDenumerators()`, it is unfortunately more complicated :-(
The problem is that if `type==0`, `type==1` or `type==2` you create new objects for `num` and `den`; while if `type==3` you return pointers to existing objects.
I don't think there is a general way of handling this, short of using `shared_ptr<TH1F>` which is probably overkill.

@dmitrijus , do you have any suggestions ?

**Comment by dmitrijus** [comment]
@fwyzard @natalia-korneeva 

Not `shared_ptr`, but a `unique_ptr<TH1>` would be fine here.
Or, perhaps, a well placed lambda.... `GetNumDenumerators(..., std::function<void(const TH1&, const TH1&)> f)`

With this you are not forced to deallocate **after** `GetNumDenumerators` exits. The only way to access these points would via this lambda. Their lifetime is limited to `GetNumDenumerators`  and this is where they should be freed. It also saves unnecessary allocations in some cases.

**Comment by fwyzard** [comment]
Sorry, but I do not understand how to make it work with `unique_ptr` :-(

[`GetNumDenumerators(...)`](https://github.com/cms-sw/cmssw/blob/master/HLTriggerOffline/Btag/src/HLTBTagHarvestingAnalyzer.cc#L125) returns pointers to the numerator and denominator via the `ptrnum` and `ptrden` parameters (but it could return those pointers by value, the result would be the same).

When `type` is 0, 1 or 2, those pointers are new TH1F objects, that later will need to be deleted.
When `type` is 3, those pointers are existing TH1F objects, that should *not* be deleted.

If you return them via `unique_ptr` they will always be deleted, which would likely crash the job after calling `GetNumDenumerators(...)` with `type` equal 3.

(you cannot return a non-deleting `unique_ptr`, because the deleter is part of the type)

One could return a copy of the TH1F also when `type` is 3: that would make the method consistent, as the returned histograms would  always need to be deleted; but it would be a waste of resources.

Indeed you could restructure the code to pass the function that will use the numerator and denominator into the equivalent of  `GetNumDenumerators(...)`, but you would still need to do something like
```c++
if (type != 3) {
  delete ptrnum;
  delete ptrden;
}
```
So you might as well do that in `dqmEndJob`, after calling `GetNumDenumerators(...)` and using the histograms.

Bottom line for @natalia-korneeva:
when you call `GetNumDenumerators(...)` with `type` other than 3, can you make sure to `delete` `ptrnum` and `ptrden` ?

**Comment by dmitrijus** [comment]
> If you return them via unique_ptr they will always be deleted, which would likely crash the job after calling GetNumDenumerators(...) with type equal 3.

Correct. This is why I suggested to use the lambda approach. This way the cleanup is not the caller's responsibility.

Returning a pointer which **may or may not** be cleaned is just asking for trouble... This is bigger mess than root, at least root guarantees that a pointer returned by Clone() will need to be cleared.
